### ISSUE TITLE
D interface files are not passed as source file

### ISSUE THREAD

**Issue by BBasile** [D-Programming-Language/dub#773]
For example if a __.di__ stands in `source/` it won't be passed as the other source files.
This case happens when one links an object file with `dflag` and that the object content is interfaced with a D interface.

Example:

http://code.dlang.org/packages/dbeaengine

In this package I had to add explicitly the di file, even if it stands in a folder specified in `sourcePaths`.

**Comment by BBasile** [comment]
Does anyone knows where d sources are filtered ?

The only results I get seem unrelated:

```
/dub/compilers/compiler.d(401,48): "auto path = getTempFile("dub_platform_probe", ".d");"
/dub/generators/visuald.d(149,56): "if (!buildsettings.sourceFiles.any!(f => f.endsWith(".d"))())"
/dub/generators/build.d(199,82): "buildsettings.sourceFiles = buildsettings.sourceFiles.filter!(s => !s.endsWith(".d"))().array();"
/dub/generators/build.d(535,63): "foreach (f; ["source/app.d", "src/app.d", "source/"~prj.name~".d", "src/"~prj.name~".d"])"
/dub/generators/build.d(535,85): "foreach (f; ["source/app.d", "src/app.d", "source/"~prj.name~".d", "src/"~prj.name~".d"])"
/dub/dub.d(386,63): "custommodname = custom_main_file.head.toString().baseName(".d");"
/dub/dub.d(391,23): "if (file.endsWith(".d") && Path(file).head.toString() != "package.d")"
/dub/dub.d(396,49): "Path mainfile = getTempFile("dub_test_root", ".d");"
/dub/dub.d(892,26): "else ret ~= p.baseName(".d");"
```

**Comment by s-ludwig** [comment]
This is the relevant place: https://github.com/D-Programming-Language/dub/blob/c5935829500dac1b6a2e5aa0dd628a3a694ab75c/source/dub/recipe/packagerecipe.d#L200

It would be interesting to clarify how this is supposed to behave from a spec point of view. Usually every .di file has a corresponding .d/.obj file which is linked together with the application somehow, so that it is not necessary, or even harmful (multiple definition linker errors), to also compile/link the .di file. If you have original content in the .di file, it should probably instead just get the .d extension, since .di files are not even required to be supported by the compiler at all (see [here](https://dlang.org/dmd-windows.html#interface-files)).

Note that .di files can also be included using a glob expression, if many are present: `"sourceFiles": ["interface/*.di"]`

**Comment by BBasile** [comment]
my point of view is that if a _.di_ stands in one of the _source_, _src_ or _sourcePaths_ folders then it should be included automatically. It's  a rare case but if someone puts it intentionally in one of these folder then he certainly wants it to be passed to the compiler.

Now the fact is that DUB doesn't allow to pass object or static library files directly (https://github.com/D-Programming-Language/dub/issues/591) and that _.di_ files always go in pair with something like that.

In short the specification would be simple: 

> if a D interface file resides in one of the X,Y,Z folders then it'll be passed to the compiler. In case of name conflict, only the _.d_ file will be passed.
